src/nnet3/nnet-simple-component.h

```diff
@@ -714,6 +714,19 @@ class NaturalGradientAffineComponent: public AffineComponent {
   // in Info().
   double max_change_scale_stats_;
 
+  /// The following function is only called if max_change_per_sample_ > 0, it returns a
+  /// scaling factor alpha <= 1.0 (1.0 in the normal case) that enforces the
+  /// "max-change" constraint.  "in_products" is the inner product with itself
+  /// of each row of the matrix of preconditioned input features; "out_products"
+  /// is the same for the output derivatives.  gamma_prod is a product of two
+  /// scalars that are output by the preconditioning code (for the input and
+  /// output), which we will need to multiply into the learning rate.
+  /// out_products is a pointer because we modify it in-place.
```

Comment:
I think there is something in the natural-gradient paper, in an appendix.

Basically the idea of the max-change is to limit the l2 norm of the
parameter change per minibatch, by scaling down that change.

However, the way it's been implemented with the natural gradient update is
a little approximate, and also rather complex-- it was done in order to
avoid the need for a temporary matrix.  I'm actually not so happy with
that, so it would be good to avoid resurrecting that code.

Actually, something just occurred to me.
Right now the way we implement the max-change for the whole model is not in
the components, but in the trainer code, where we anyway have to compute
the individual l2 norms for the components, as part of computing the
whole-model change.
I think it would be better, rather than resurrecting the old, nasty code
for max-change, to implement it at that level.
By means of dynamic casting etc., we could determine which components
inherit from UpdatableComponent and have a max-change set, and apply the
max-changes individually in the trainer code.
Note: in this case the max-change would really be per minibatch, and not
per matrix operation as it is when it's implemented inside the components.

Dan


On Fri, Jul 29, 2016 at 2:34 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.h
> <https://github.com/kaldi-asr/kaldi/pull/918#discussion_r72864362>:
>
> > @@ -714,6 +714,19 @@ class NaturalGradientAffineComponent: public AffineComponent {
> >    // in Info().
> >    double max_change_scale_stats_;
> >
> > +  /// The following function is only called if max_change_per_sample_ > 0, it returns a
> > +  /// scaling factor alpha <= 1.0 (1.0 in the normal case) that enforces the
> > +  /// "max-change" constraint.  "in_products" is the inner product with itself
> > +  /// of each row of the matrix of preconditioned input features; "out_products"
> > +  /// is the same for the output derivatives.  gamma_prod is a product of two
> > +  /// scalars that are output by the preconditioning code (for the input and
> > +  /// output), which we will need to multiply into the learning rate.
> > +  /// out_products is a pointer because we modify it in-place.
>
> Is it possible to provide a description of what max-change intends to do
> in detail ?
> Something like a reference to a paper or at least some equations here
> describing whats being done and why.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/918/files/6657686d5c8593f2863a2a046b07aaba88099bc2#r72864362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuwnh7tzB-kseshgqMBaBS-0SZ8-Lks5qanH-gaJpZM4JUf8N>
> .
>